File: desktop/src/taxonomy.ts
import type { BusinessDocument, BusinessIndex, CompanySummary } from './business';

export type FileRecord = { path: string; sha256: string; bytes: number };
export type DiveRecord = { folder: string; label: string; documents: FileRecord[] };
export type Sector = { id: string; name_sv: string; name_en: string; slug: string };
export type Branch = { id: string; sector_id: string; name_sv: string; name_en: string; study_path: string; study_status: 'scaffold' | 'graduated'; corpus_status: 'none' | 'graduated'; overview: FileRecord; deep_dives: DiveRecord[]; shared_study_ids: string[] };
export type Correction = { company_id: string; expected_sector_id: string; expected_branch_id: string; branch_id: string; reason: string; source: string; reviewed_at: string };
export type Classification = { company_id: string; source_sector_id: string; source_branch_id: string; sector_id: string; branch_id: string; status: 'source' | 'corrected' | 'aligned' | 'needs_review'; correction: Correction | null };
export type SharedStudy = { id: string; path: string; branch_ids: string[]; overview: FileRecord; deep_dives: DiveRecord[] };
export type Taxonomy = {
  version: number; as_of: string; exported_at: string; business_sha256: string | null; classification_as_of: string | null;
  sectors: Record<string, Sector>; branches: Record<string, Branch>; classifications: Record<string, Classification>;
  shared_studies: Record<string, SharedStudy>; unmapped_deep_dives: DiveRecord[]; sources: FileRecord[]; corrections_sha256: string; notes: string[];
};
export const fold = (text: string) => text.normalize('NFD').replace(/[\u0300-\u036f]/g, '').toLowerCase().trim();
export const companyBranch = (company: CompanySummary, taxonomy: Taxonomy | null) => taxonomy?.classifications[company.id]?.branch_id ?? company.branch_id;
export const branchCompanies = (index: BusinessIndex | null, taxonomy: Taxonomy | null, branch: string) => Object.values(index?.companies ?? {}).filter(c => companyBranch(c, taxonomy) === branch);
export function branchDives(taxonomy: Taxonomy, branch: Branch) {
  return [...new Map([...branch.deep_dives, ...branch.shared_study_ids.flatMap(id => taxonomy.shared_studies[id].deep_dives)].map(d => [d.folder, d])).values()];
}
export function directoryTotals(taxonomy: Taxonomy) {
  const dives = [...Object.values(taxonomy.branches).flatMap(b => b.deep_dives), ...Object.values(taxonomy.shared_studies).flatMap(s => s.deep_dives), ...taxonomy.unmapped_deep_dives];
  return { sectors: Object.keys(taxonomy.sectors).length, branches: Object.keys(taxonomy.branches).length, studies: Object.values(taxonomy.branches).filter(b => b.study_status === 'graduated').length, dives: new Set(dives.map(d => d.folder)).size };
}
export function findBranches(taxonomy: Taxonomy, query: string, sector: string, coverage: string, index: BusinessIndex | null) {
  const term = fold(query);
  return Object.values(taxonomy.branches).filter(b => (sector === 'all' || b.sector_id === sector)
    && fold(`${b.id} ${b.name_sv} ${b.name_en} ${taxonomy.sectors[b.sector_id].name_sv} ${taxonomy.sectors[b.sector_id].name_en}`).includes(term)
    && (coverage === 'all' || coverage === 'profiles' && branchCompanies(index, taxonomy, b.id).length > 0 || coverage === 'studies' && b.study_status === 'graduated' || coverage === 'dives' && branchDives(taxonomy, b).length > 0))
    .sort((a, b) => a.name_en.localeCompare(b.name_en));
}

export function validateTaxonomy(raw: any, business: BusinessDocument | null, businessHash: string | null): asserts raw is Taxonomy {
  const check = (ok: unknown, message = 'The taxonomy directory is invalid.') => { if (!ok) throw new Error(message); };
  const obj = (x: any) => !!x && typeof x === 'object' && !Array.isArray(x);
  const text = (x: any) => typeof x === 'string' && x.length > 0 && new TextEncoder().encode(x).length <= 20000;
  const id = (x: any) => typeof x === 'string' && /^[1-9][0-9]{0,9}$/.test(x);
  const hash = (x: any) => typeof x === 'string' && /^[a-f0-9]{64}$/.test(x);
  const path = (x: any) => text(x) && !/[\\:]/.test(x) && x.split('/').every((p: string) => p && p !== '.' && p !== '..');
  const date = (x: any) => typeof x === 'string' && /^20\d{2}-\d{2}-\d{2}$/.test(x) && Number.isFinite(Date.parse(x)) && new Date(x).toISOString().slice(0, 10) === x;
  const list = (x: any, max = 1000): any[] => { check(Array.isArray(x) && x.length <= max); return x; };
  const objects = (x: any, max: number): [string, any][] => { check(obj(x) && Object.keys(x).length <= max); return Object.entries(x); };
  const file = (x: any) => check(obj(x) && path(x.path) && hash(x.sha256) && Number.isSafeInteger(x.bytes) && x.bytes >= 0, 'Invalid taxonomy source file.');
  const dives = (rows: any) => { const seen = new Set(); for (const d of list(rows, 10000)) { check(obj(d) && path(d.folder) && text(d.label) && !seen.has(d.folder)); seen.add(d.folder); const docs = list(d.documents); check(docs.length > 0); const names = new Set(); for (const f of docs) { file(f); check(f.path.startsWith(`${d.folder}/`) && !names.has(f.path)); names.add(f.path); } } };
  check(raw?.version === 1 && date(raw.as_of), 'Unsupported taxonomy version or inventory date.');
  check(typeof raw.exported_at === 'string' && /^20\d{2}-\d{2}-\d{2}T(?:[01]\d|2[0-3]):[0-5]\d:[0-5]\d(?:\.\d+)?(?:Z|\+00:00)$/.test(raw.exported_at) && date(raw.exported_at.slice(0, 10)) && Number.isFinite(Date.parse(raw.exported_at)));
  const sectors = objects(raw.sectors, 100); check(sectors.length > 0);
  for (const [key, s] of sectors) check(obj(s) && id(key) && s.id === key && text(s.name_sv) && text(s.name_en) && typeof s.slug === 'string' && /^[a-z0-9]+(?:-[a-z0-9]+)*$/.test(s.slug));
  const branches = objects(raw.branches, 1000); check(branches.length > 0);
  for (const [key, b] of branches) { check(obj(b) && id(key) && b.id === key && id(b.sector_id) && raw.sectors[b.sector_id] && text(b.name_sv) && text(b.name_en) && path(b.study_path) && ['scaffold', 'graduated'].includes(b.study_status) && ['none', 'graduated'].includes(b.corpus_status)); file(b.overview); check(b.overview.path === `${b.study_path}/README.md`); dives(b.deep_dives); check(list(b.shared_study_ids).every(text)); }
  const groups = objects(raw.shared_studies, 1000);
  for (const [key, g] of groups) { check(obj(g) && /^[a-z][a-z0-9_]{0,80}$/.test(key) && g.id === key && path(g.path)); file(g.overview); check(g.overview.path.startsWith(`studies/sectors/${g.path}/`)); const members = list(g.branch_ids); check(members.length >= 2 && new Set(members).size === members.length && members.every((bid: string) => id(bid) && raw.branches[bid]?.shared_study_ids.includes(key))); dives(g.deep_dives); }
  for (const [key, b] of branches) check(new Set(b.shared_study_ids).size === b.shared_study_ids.length && b.shared_study_ids.every((group: string) => raw.shared_studies[group]?.branch_ids.includes(key)), 'Invalid shared taxonomy study membership.');
  dives(raw.unmapped_deep_dives); for (const f of list(raw.sources)) file(f); check(list(raw.notes).every(text) && hash(raw.corrections_sha256));
  check(raw.business_sha256 === businessHash && raw.classification_as_of === (business?.as_of ?? null), 'Taxonomy does not match the selected business document.');
  const assignments = objects(raw.classifications, 300);
  check(assignments.length === Object.keys(business?.companies ?? {}).length, 'Taxonomy company coverage does not match the business document.');
  for (const [key, c] of assignments) {
    const original = business?.companies[key];
    check(obj(c) && id(key) && c.company_id === key && original && c.source_sector_id === original.sector_id && c.source_branch_id === original.branch_id && raw.branches[c.source_branch_id]?.sector_id === c.source_sector_id, 'Taxonomy original classification does not match the business document.');
    let status = 'source', target = c.source_branch_id;
    if (c.correction !== null) {
      const r = c.correction;
      check(obj(r) && r.company_id === key && id(r.expected_branch_id) && id(r.expected_sector_id) && raw.branches[r.expected_branch_id]?.sector_id === r.expected_sector_id && id(r.branch_id) && raw.branches[r.branch_id] && text(r.reason) && r.reason.trim() && text(r.source) && r.source.trim() && date(r.reviewed_at) && r.reviewed_at <= raw.as_of, 'Invalid reviewed taxonomy correction.');
      if (c.source_branch_id === r.branch_id) status = 'aligned';
      else if (c.source_branch_id === r.expected_branch_id && c.source_sector_id === r.expected_sector_id) { status = 'corrected'; target = r.branch_id; }
      else status = 'needs_review';
    }
    check(c.status === status && c.branch_id === target && c.sector_id === raw.branches[target]?.sector_id, 'Taxonomy effective classification disagrees with its correction.');
  }
}

File: desktop/research-store/src/taxonomy.rs
//! Validated directory metadata and reversible, source-bound classification corrections.
use super::business::company_id;
use super::{array, check, date, fields, is_id, object, timestamp, Result};
use serde_json::Value;
use std::collections::BTreeSet;

fn text(v: &Value) -> bool {
    v.as_str()
        .is_some_and(|s| !s.is_empty() && s.len() <= 20000)
}
fn identifier(v: &Value) -> bool {
    v.as_str().is_some_and(company_id)
}
fn day(v: &Value) -> bool {
    v.as_str().is_some_and(|s| s.starts_with("20") && date(s))
}
fn path(v: &Value) -> bool {
    text(v)
        && v.as_str().is_some_and(|s| {
            !s.contains(['\\', ':']) && s.split('/').all(|p| !p.is_empty() && p != "." && p != "..")
        })
}
fn file(v: &Value) -> Result<()> {
    check(
        path(&v["path"])
            && v["sha256"].as_str().is_some_and(is_id)
            && v["bytes"].as_u64().is_some_and(|n| n <= 9007199254740991),
        "Invalid taxonomy source file.",
    )
}
fn dives(v: &Value) -> Result<()> {
    let rows = array(v)?;
    check(rows.len() <= 10000, "Too many taxonomy deep dives.")?;
    let mut seen = BTreeSet::new();
    for row in rows {
        check(
            path(&row["folder"]) && text(&row["label"]),
            "Invalid taxonomy deep-dive metadata.",
        )?;
        let folder = row["folder"].as_str().unwrap();
        check(seen.insert(folder), "Duplicate taxonomy deep-dive folder.")?;
        let docs = array(&row["documents"])?;
        check(
            !docs.is_empty() && docs.len() <= 1000,
            "Invalid deep-dive document inventory.",
        )?;
        let mut paths = BTreeSet::new();
        for doc in docs {
            file(doc)?;
            let p = doc["path"].as_str().unwrap();
            check(
                p.starts_with(&format!("{folder}/")) && paths.insert(p),
                "Invalid or duplicate deep-dive source path.",
            )?;
        }
    }
    Ok(())
}
pub(super) fn export_timestamp(value: &Value) -> bool {
    timestamp(value)
        && value.as_str().is_some_and(|s| {
            let local = s
                .strip_suffix('Z')
                .or_else(|| s.strip_suffix("+00:00"))
                .unwrap();
            let fraction = &local[19..];
            s.starts_with("20")
                && (fraction.is_empty()
                    || fraction.starts_with('.')
                        && fraction.len() > 1
                        && fraction[1..].bytes().all(|b| b.is_ascii_digit()))
        })
}
pub fn validate(raw: &Value, business: Option<&Value>, business_hash: Option<&str>) -> Result<()> {
    check(
        (raw["version"] == 1 || raw["version"] == 2)
            && day(&raw["as_of"])
            && export_timestamp(&raw["exported_at"]),
        "Unsupported taxonomy version or inventory date.",
    )?;
    let sectors = object(&raw["sectors"])?;
    check(
        !sectors.is_empty() && sectors.len() <= 100,
        "Invalid taxonomy sectors.",
    )?;
    for (key, s) in sectors {
        check(
            company_id(key)
                && s["id"] == *key
                && text(&s["name_sv"])
                && text(&s["name_en"])
                && s["slug"].as_str().is_some_and(|slug| {
                    !slug.is_empty()
                        && slug.split('-').all(|p| {
                            !p.is_empty()
                                && p.bytes()
                                    .all(|b| b.is_ascii_lowercase() || b.is_ascii_digit())
                        })
                }),
            "Invalid taxonomy sector metadata.",
        )?;
    }
    let branches = object(&raw["branches"])?;
    check(
        !branches.is_empty() && branches.len() <= 1000,
        "Invalid taxonomy branches.",
    )?;
    for (key, b) in branches {
        check(
            company_id(key)
                && b["id"] == *key
                && identifier(&b["sector_id"])
                && sectors.contains_key(b["sector_id"].as_str().unwrap())
                && text(&b["name_sv"])
                && text(&b["name_en"])
                && path(&b["study_path"])
                && ["scaffold", "graduated"].contains(&b["study_status"].as_str().unwrap_or(""))
                && ["none", "graduated"].contains(&b["corpus_status"].as_str().unwrap_or("")),
            "Invalid taxonomy branch metadata.",
        )?;
        file(&b["overview"])?;
        check(
            b["overview"]["path"] == format!("{}/README.md", b["study_path"].as_str().unwrap()),
            "Invalid branch overview path.",
        )?;
        dives(&b["deep_dives"])?;
        let shared = array(&b["shared_study_ids"])?;
        check(
            shared.len() <= 1000 && shared.iter().all(text),
            "Invalid shared study identifiers.",
        )?;
    }
    let groups = object(&raw["shared_studies"])?;
    check(groups.len() <= 1000, "Too many shared studies.")?;
    for (key, g) in groups {
        check(
            !key.is_empty()
                && key.len() <= 81
                && key.as_bytes()[0].is_ascii_lowercase()
                && key
                    .bytes()
                    .all(|c| c.is_ascii_lowercase() || c.is_ascii_digit() || c == b'_')
                && g["id"] == *key
                && path(&g["path"]),
            "Invalid shared study metadata.",
        )?;
        file(&g["overview"])?;
        check(
            g["overview"]["path"]
                .as_str()
                .unwrap()
                .starts_with(&format!("studies/sectors/{}/", g["path"].as_str().unwrap())),
            "Invalid shared study overview.",
        )?;
        let members = array(&g["branch_ids"])?;
        check(
            members.len() >= 2 && members.len() <= 1000,
            "Invalid shared study membership.",
        )?;
        let mut seen = BTreeSet::new();
        for member in members {
            check(identifier(member), "Invalid shared study branch.")?;
            let bid = member.as_str().unwrap();
            check(
                seen.insert(bid)
                    && branches.get(bid).is_some_and(|b| {
                        b["shared_study_ids"]
                            .as_array()
                            .unwrap()
                            .iter()
                            .any(|id| id == key)
                    }),
                "Invalid shared taxonomy study membership.",
            )?;
        }
        dives(&g["deep_dives"])?;
    }
    for (key, b) in branches {
        let mut seen = BTreeSet::new();
        for id in b["shared_study_ids"].as_array().unwrap() {
            let id = id.as_str().unwrap();
            check(
                seen.insert(id)
                    && groups.get(id).is_some_and(|g| {
                        g["branch_ids"]
                            .as_array()
                            .unwrap()
                            .iter()
                            .any(|bid| bid == key)
                    }),
                "Invalid shared taxonomy study membership.",
            )?;
        }
    }
    dives(&raw["unmapped_deep_dives"])?;
    let sources = array(&raw["sources"])?;
    check(sources.len() <= 1000, "Too many taxonomy sources.")?;
    for source in sources {
        file(source)?;
    }
    let notes = array(&raw["notes"])?;
    check(
        notes.len() <= 1000
            && notes.iter().all(text)
            && raw["corrections_sha256"].as_str().is_some_and(is_id),
        "Invalid taxonomy notes or correction checksum.",
    )?;
    let with_catalogue = raw["version"] == 2;
    if with_catalogue {
        super::catalogue::validate(&raw["catalogue"], business, raw["as_of"].as_str().unwrap())?;
    } else {
        check(
            raw.get("catalogue").is_none(),
            "A v1 taxonomy cannot include an unvalidated company catalogue.",
        )?;
    }
    let expected_hash = business_hash.map_or(Value::Null, Value::from);
    let expected_date = if with_catalogue {
        raw["catalogue"]["as_of"].clone()
    } else {
        business.map_or(Value::Null, |b| b["as_of"].clone())
    };
    check(
        raw.get("business_sha256") == Some(&expected_hash)
            && raw.get("classification_as_of") == Some(&expected_date),
        "Taxonomy does not match the selected business document.",
    )?;
    let classifications = object(&raw["classifications"])?;
    let companies = if with_catalogue {
        Some(object(&raw["catalogue"]["listings"])?)
    } else {
        business.and_then(|b| b["companies"].as_object())
    };
    check(
        classifications.len() <= if with_catalogue { 100000 } else { 300 }
            && classifications.len() == companies.map_or(0, |c| c.len()),
        "Taxonomy company coverage does not match the selected directory.",
    )?;
    for (key, c) in classifications {
        fields(c, &["company_id", "status"])?;
        for k in [
            "source_sector_id",
            "source_branch_id",
            "sector_id",
            "branch_id",
        ] {
            check(
                c.get(k).is_some() && (c[k].is_null() || identifier(&c[k])),
                "Invalid classification identifier.",
            )?;
        }
        check(
            company_id(key)
                && c["company_id"] == *key
                && companies.and_then(|c| c.get(key)).is_some_and(|original| {
                    c["source_sector_id"] == original["sector_id"]
                        && c["source_branch_id"] == original["branch_id"]
                }),
            "Taxonomy original classification does not match its source document.",
        )?;
        let source_branch = c["source_branch_id"].as_str();
        let source_parent = source_branch.and_then(|id| branches.get(id));
        if !with_catalogue {
            check(
                source_parent.is_some_and(|b| b["sector_id"] == c["source_sector_id"]),
                "Invalid original branch parent.",
            )?;
        }
        let mut target = source_branch.filter(|id| branches.contains_key(*id));
        let mut status = if target.is_none() {
            "unclassified"
        } else if source_parent.unwrap()["sector_id"] == c["source_sector_id"] {
            "source"
        } else {
            "sector_mismatch"
        };
        check(
            c.get("correction").is_some(),
            "Missing taxonomy correction state.",
        )?;
        let r = &c["correction"];
        if !r.is_null() {
            fields(
                r,
                &[
                    "company_id",
                    "expected_sector_id",
                    "expected_branch_id",
                    "branch_id",
                    "reason",
                    "source",
                ],
            )?;
            check(
                r["company_id"] == *key
                    && identifier(&r["expected_branch_id"])
                    && identifier(&r["expected_sector_id"])
                    && identifier(&r["branch_id"])
                    && branches.contains_key(r["branch_id"].as_str().unwrap())
                    && !r["reason"].as_str().unwrap().trim().is_empty()
                    && !r["source"].as_str().unwrap().trim().is_empty()
                    && day(&r["reviewed_at"])
                    && r["reviewed_at"].as_str() <= raw["as_of"].as_str(),
                "Invalid reviewed taxonomy correction.",
            )?;
            if !with_catalogue {
                check(
                    branches
                        .get(r["expected_branch_id"].as_str().unwrap())
                        .is_some_and(|b| b["sector_id"] == r["expected_sector_id"]),
                    "Invalid correction original branch parent.",
                )?;
            }
            if c["source_branch_id"] == r["branch_id"]
                && source_parent.is_some_and(|b| b["sector_id"] == c["source_sector_id"])
            {
                status = "aligned";
            } else if c["source_branch_id"] == r["expected_branch_id"]
                && c["source_sector_id"] == r["expected_sector_id"]
            {
                status = "corrected";
                target = r["branch_id"].as_str();
            } else {
                status = "needs_review";
            }
        }
        let expected_branch = target.map_or(Value::Null, Value::from);
        let expected_sector = target.map_or(&Value::Null, |id| &branches[id]["sector_id"]);
        check(
            c["status"] == status
                && c["branch_id"] == expected_branch
                && &c["sector_id"] == expected_sector,
            "Taxonomy effective classification disagrees with its correction.",
        )?;
    }
    Ok(())
}
